lua/lazyvim/plugins/extras/lang/java.lua

```diff
@@ -1,125 +1,143 @@
-                    vim.keymap.set("n", "<leader>tr", function() require("jdtls").test_nearest_method({ bufnr = buffer }) end, { buffer = buffer, desc = "Run nearest" })
-                  end
-                  if mason_registry.has_package("java-debug-adapter") then
-                    -- custom init for Java debugger
-                    require("jdtls").setup_dap({ hotcodereplace = "auto" })
-                    require("jdtls.dap").setup_dap_main_class_configs()
-                  end
-                  require("jdtls.setup").add_commands()
-                end),
-                cmd = cmd,
-                root_dir = root_dir,
-                init_options = {
-                  bundles = bundles,
-                }
-              }
-              local jdtls_opts = require("lazyvim.util").opts("nvim-jdtls")
-              require("jdtls").start_or_attach(vim.tbl_deep_extend("force", jdtls_opts or {}, jdtls_base_config))
-              require("which-key").register({ c = { x = { name = "Extract" } } }, { prefix = "<leader>" })
-            end,
-          })
-          return true -- avoid duplicate servers
+          return true
         end,
       },
     },
   },
+  {
+    "mfussenegger/nvim-jdtls",
+    dependencies = {
+      "neovim/nvim-lspconfig",
+      "mfussenegger/nvim-dap",
+      "hrsh7th/nvim-cmp",
+      "folke/which-key.nvim",
+    },
+    ft = { "java" },
+    config = function()
+      local project_name = vim.fn.fnamemodify(vim.fn.getcwd(), ":p:h:t")
+      local workspace_dir = vim.fn.stdpath("data") .. "/lsp/java/" .. project_name
+      local config = {
+        cmd = {
+          "java",
+          "-Declipse.application=org.eclipse.jdt.ls.core.id1",
+          "-Dosgi.bundles.defaultStartLevel=4",
+          "-Declipse.product=org.eclipse.jdt.ls.core.product",
+          "-Dlog.protocol=true",
+          "-Dlog.level=ALL",
+          "-Xmx1g",
+          "--add-modules=ALL-SYSTEM",
+          "--add-opens",
+          "java.base/java.util=ALL-UNNAMED",
+          "--add-opens",
+          "java.base/java.lang=ALL-UNNAMED",
+          "-jar",
+          vim.fn.expand("$MASON/share/jdtls/plugins/org.eclipse.equinox.launcher.jar"),
+          "-configuration",
+          vim.fn.expand("$MASON/share/jdtls/config"),
+          "-data",
+          workspace_dir,
+        },
+        root_dir = require("jdtls.setup").find_root({
+          ".git",
+          "mvnw",
+          "gradlew",
+          "pom.xml",
+          "build.gradle",
+          "build.gradle.kts",
+          "settings.gradle",
+          "settings.gradle.kts",
+          "build.xml",
+          ".project",
+        }),
+        settings = {
+          java = {},
+        },
+        handlers = {
+          ["$/progress"] = function() end,
+          ["language/status"] = function() end,
+        },
+        init_options = {
+          bundles = {
+            vim.fn.expand("$MASON/share/java-debug-adapter/com.microsoft.java.debug.plugin.jar"),
```

Comment:
The original version checks if the user has installed the debug adapter and test extension… you assume here it’s always there…

Reply:
I want to provide a full feature config (not only include LSP, but also DAP, keymaps, and Test support), and I have added the debug adapter and test extension in the `mason-nvim-dap.nvim` ensure_installed, So I think they should always be there...

For instead of referring $MASON, sorry for that, I didn't know this before, my bad